src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs

```diff
@@ -71,16 +78,21 @@ internal HttpRetryHandler(IEnvironmentVariableReader environmentVariableReader)
             }
 
             var tries = 0;
-            HttpResponseMessage response = null;
+            HttpResponseMessage? response = null;
             var success = false;
+            TimeSpan? retryAfter = null;
 
             while (tries < request.MaxTries && !success)
             {
                 // There are many places where another variable named "MaxTries" is set to 1,
                 // so the Delay() never actually occurs.
                 // When opted in to "enhanced retry", do the delay and have it increase exponentially where applicable
                 // (i.e. when "tries" is allowed to be > 1)
-                if (tries > 0 || (_enhancedHttpRetryHelper.IsEnabled && request.IsRetry))
+                if (retryAfter != null && _enhancedHttpRetryHelper.ObserveRetryAfter)
+                {
+                    await Task.Delay(retryAfter.Value, cancellationToken);
```

Comment:
I have not looked at the entire PR more closely, but I think we may have to set `retryAfter = null` after line no 93. The reason is that if we end up retrying the request again then we don't want to add additional delay because of previous 429 response.

Here is my understanding.
1. NuGet tries a request, but the server responds with 429 response.
2. NuGet waits based on `Retry-After` value before retrying the request.
3. Let's say server responds with 503 response then the current logic delays sending the request because `retryAfter` will not be in this case.

Reply:
It will get set to null again on line 212